src/mawpy/steps/incremental_clustering.py

```diff
@@ -0,0 +1,396 @@
+"""
+Perform clustering on multiple locations of one user based on a spatial threshold
+Each cluster of locations represents a potential stay
+
+input:
+    gps stay information / celluar stay information
+    spatial threshold
+    duration constraint threshold (for detect common stay)
+output:
+    potential stays represented by clusters of locations
+"""
+
+import sys
+from datetime import datetime
+
+import psutil
+import time
+import numpy as np
+import pandas as pd
+
+from multiprocessing import Pool
+from multiprocessing import Lock, cpu_count
+
+from geopy.distance import distance
+from sklearn.cluster import KMeans
+
+from mawpy.cluster import Cluster
+from mawpy.constants import USER_ID, STAY_DUR, ORIG_LAT, STAY_LAT, STAY_LONG, UNIX_START_T, UNIX_START_DATE, STAY_UNC, \
+    ORIG_LONG, ORIG_UNC, STAY, STAY_LAT_PRE_COMBINED, STAY_LONG_PRE_COMBINED, STAY_PRE_COMBINED
+from mawpy.utilities.preprocessing import get_preprocessed_dataframe, get_list_of_chunks_by_column
+
+
+def init(this_lock):
+    global lock
+    lock = this_lock
+
+
+def _get_cluster_center(row, mapping, dur_constr):
```

Comment:
Since this whole module is getting re-written it might be a good opportunity to add type hints? Or is there a reason to defer this?

Reply:
Yeah, I will add this going forward in the next PR.